### PR TITLE
feat: added base http data type

### PR DESCRIPTION
**Pull request by cl-shaharshavit-rnd**

### CODE REVIEW

File: MergerLogic/Utils/httpUtils.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using MergerLogic.Batching;
+
+namespace MergerLogic.Utils
+{
+    internal class httpUtils : DataUtils
+    {
+        private PathPatternUtils _pathPatternUtils;
```
Comment (shimoncohen) [inline MergerLogic/Utils/httpUtils.cs:7]: Convention is using `mPropertyName`.

File: MergerLogic/Utils/httpUtils.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            resTask.Wait();
+            var httpRes = resTask.Result;
+            if (httpRes.StatusCode == System.Net.HttpStatusCode.NotFound)
+            {
+                return null;
+            }
+            var bodyTask = httpRes.Content.ReadAsByteArrayAsync();
+            byte[] data = bodyTask.Result;
+            return new Tile(z, x, y, data);
+
```
Comment (shimoncohen) [inline MergerLogic/Utils/httpUtils.cs:28]: Remove empty line.

File: MergerLogic/Utils/httpUtils.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            string url = this._pathPatternUtils.renderUrlTemplate(x, y, z);
+            var resTask = this._httpClient.GetAsync(url);
+            resTask.Wait();
+            var httpRes = resTask.Result;
+            if (httpRes.StatusCode == System.Net.HttpStatusCode.NotFound)
+            {
+                return null;
+            }
+            var bodyTask = httpRes.Content.ReadAsByteArrayAsync();
```
Comment (shimoncohen) [inline MergerLogic/Utils/httpUtils.cs:25]: No need to wait?

File: MergerLogic/Utils/PathPatternUtils.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+        public PathPatternUtils(string pattern)
+        {
+            this.parts = new string[4];
+            this.keys = new string[3];
+            this.keyValues = new Dictionary<string, string>(9);
+            this.compilePattern(pattern);
+        }
+
+        private void compilePattern(string pattern)
```
Comment (shimoncohen) [inline MergerLogic/Utils/PathPatternUtils.cs:24]: Why not just split by {x} and so on depending on the source type?

File: MergerLogic/Batching/Tile.cs [hunk 5]
```diff
@@ -16,19 +22,22 @@ public int X
         public int Y
         { get; private set; }
 
-        public int BlobSize
-        { get; private set; }
+        private byte[] _data;
```
Comment (shimoncohen) [inline MergerLogic/Batching/Tile.cs:25]: Convention is `mPropertyName`.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerLogic/Utils/httpUtils.cs (hunk 1 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/httpUtils.cs:7]: ???

File: MergerLogic/Utils/PathPatternUtils.cs (hunk 4 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/PathPatternUtils.cs:19]: this is more generic then multiple splits (order of Parma's don't matter) and should be faster too

File: MergerLogic/Utils/httpUtils.cs (hunk 3 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/httpUtils.cs:25]: fixed

File: MergerLogic/Utils/httpUtils.cs (hunk 1 above)
Comment (asafMasa) [reply MergerLogic/Utils/httpUtils.cs:7]: use this conventions
https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions

File: MergerLogic/Utils/PathPatternUtils.cs (hunk 4 above)
Comment (asafMasa) [reply MergerLogic/Utils/PathPatternUtils.cs:19]: i want to see benchmark tests for:
String.Format
StringBuilder
String.replace
Regex
Current existing code in PR

For Build and Render

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: MergerLogic/DataTypes/TMS.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿using MergerLogic.Batching;
+
+namespace MergerLogic.DataTypes
+{
+    public class TMS : HttpDataSource
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/TMS.cs:5]: all other sources use that convention

File: MergerLogic/DataTypes/WMTS.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using MergerLogic.Batching;
+
+namespace MergerLogic.DataTypes
+{
+    public class WMTS : HttpDataSource
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/WMTS.cs:5]: all other sources use that convention

File: MergerLogic/DataTypes/XYZ.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿using MergerLogic.Batching;
+
+namespace MergerLogic.DataTypes
+{
+    public class XYZ : HttpDataSource
+    {
+        public XYZ(DataType type, string path, int batchSize, Extent extent, int maxZoom, int minZoom = 0)
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/XYZ.cs:7]: all other sources use that convention

File: MergerLogic/DataTypes/HttpDataSource.cs [hunk 9]
```diff
@@ -1,43 +1,38 @@
...
         protected TileRange[] tileRanges;
         protected IEnumerator<Tile[]> batches;
         protected int batchIndex = 0;
-        protected HttpDataSource(DataType type, string path, int batchSize,Extent extent, TileGridOrigin origin, int maxZoom, int minZoom = 0) : base(type, path, batchSize, null)
+        protected HttpDataSource(DataType type, string path, int batchSize, Extent extent, TileGridOrigin origin, int maxZoom, int minZoom = 0) : base(type, path, batchSize, null)
         {
             var patternUtils = new PathPatternUtils(path);
             this.utils = new httpUtils(path, patternUtils);
-            GenTilesRanges(extent, origin, minZoom, maxZoom);
+            this.GenTilesRanges(extent, origin, minZoom, maxZoom);
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/HttpDataSource.cs:15]: Rename to `GenTileRanges`.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerCli/Program.cs [hunk 10]
```diff
@@ -1,4 +1,5 @@
 ﻿using MergerCli.Utils;
```
Comment (shimoncohen) [inline MergerCli/Program.cs:1]: Add log indicating a resume, and some info on the resume operation.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerCli/Program.cs (hunk 10 above)
Comment (cl-shaharshavit-rnd) [reply MergerCli/Program.cs:1]: added

File: MergerLogic/DataTypes/HttpDataSource.cs (hunk 9 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/HttpDataSource.cs:15]: fixed

File: MergerLogic/Utils/PathPatternUtils.cs [hunk 11]
```diff
@@ -1,69 +1,30 @@
...
         private void compilePattern(string pattern)
         {
-            int lastPartEnd = 0;
-            int partIndx = 0;
-            int keyIdx = 0;
-            bool isOpenedVariable = false;
-            
-            for(int i = 0; i < pattern.Length; i++)
+            this._pattern = Regex.Split(pattern, "{(x)}|{(X)}|{(TileCol)}|{(y)}|{(Y)}|{(TileRow)}|{(TileMatrix)}|{(z)}|{(Z)}");
+            if (this._pattern.Length == 6)
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/PathPatternUtils.cs:20]: added

**Review by shimoncohen — APPROVED** [review]

**Pull request merged**